### PR TITLE
Support pyarrow output for GenomicsDB queries

### PR DESCRIPTION
**Pull request by nalinigans**

### CODE REVIEW

File: test/test_genomicsdb.py [hunk 1]
```diff
@@ -10,160 +11,243 @@
...
-  if not os.path.exists(tmp_dir):
-    os.makedirs(tmp_dir)
-  else:
-    sys.exit("Aborting as temporary directory seems to exist!")
-  tar = tarfile.open("test/inputs/sanity.test.tgz")
-  tar.extractall(tmp_dir)
-  os.chdir(tmp_dir)
-  yield
-  shutil.rmtree(tmp_dir)
+    tmp_dir = tempfile.TemporaryDirectory().name
```
Comment (nalinigans) [inline test/test_genomicsdb.py:17]: Sorry, this got reformatted automatically by Emacs. I can bring it back to where it was before if it is too difficult to read. But the arrow changes are from L108-L131!

File: .github/scripts/install_prereqs.sh [hunk 2]
```diff
@@ -195,15 +203,15 @@ if [[ $1 == "release" ]]; then
   echo "DYLD_LIBRARY_PATH=${DYLD_LIBRARY_PATH}"
   echo "LD_LIBRARY_PATH=${LD_LIBRARY_PATH}"
   # For Debugging
-  git clone https://github.com/GenomicsDB/GenomicsDB.git -b develop $NATIVE_BUILD_DIR
+  git clone https://github.com/GenomicsDB/GenomicsDB.git -b ng_nanoarrow $NATIVE_BUILD_DIR
```
Comment (nalinigans) [inline .github/scripts/install_prereqs.sh:206]: Will revert to develop once https://github.com/GenomicsDB/GenomicsDB/pull/338 is approved.

File: .github/workflows/basic.yml [hunk 3]
```diff
@@ -15,12 +15,12 @@ on:
 env:
   GENOMICSDB_BUILD_DIR: ${{github.workspace}}
   GENOMICSDB_HOME: ${{github.workspace}}/install
-  GENOMICSDB_BRANCH: develop
+  GENOMICSDB_BRANCH: ng_nanoarrow
```
Comment (nalinigans) [inline .github/workflows/basic.yml:18]: Will revert to develop once https://github.com/GenomicsDB/GenomicsDB/pull/338 is approved.

File: src/genomicsdb.pyx [hunk 4]
```diff
@@ -305,6 +312,55 @@ cdef class _GenomicsDB:
...
+        processor.set_threaded(1)
+
+      def query_calls():
+        array_name = as_string(array)
+        scan_range = scan_full()
+        with nogil:
+          self._genomicsdb.query_variant_calls(processor, array_name, scan_range)
+
+      if non_blocking:
+        query_thread = threading.Thread(target=query_calls)
```
Comment (nalinigans) [inline src/genomicsdb.pyx:338]: Will be changing this to using the thread pool with asyncio

File: requirements.txt [hunk 5]
```diff
@@ -28,3 +28,4 @@
 numpy>=1.24.0
 pandas>=2.1.0
 protobuf>=4.21.1
+pyarrow
```
Comment (oliviawebber) [inline requirements.txt:31]: Could a version specifier be added to this dependency?

File: src/genomicsdb.pyx [hunk 6]
```diff
@@ -305,6 +312,78 @@ cdef class _GenomicsDB:
...
+        query_thread.start()
+      else:
+        query_calls()
+
+      cdef object schema_capsule
+      while True:
+        arrow_array = processor.arrow_array()
+        if arrow_schema == NULL:
```
Comment (oliviawebber) [inline src/genomicsdb.pyx:369]: Is this first block only to initialize and yield the schema on the first iteration? If so, could it be moved out of the `while` block?

File: src/genomicsdb.pyx [hunk 7]
```diff
@@ -305,6 +312,78 @@ cdef class _GenomicsDB:
...
+          arrow_schema = processor.arrow_schema();
+          schema_capsule = pycapsule_get_arrow_schema(arrow_schema)
+          schema_obj = _ArrowSchemaWrapper._import_from_c_capsule(schema_capsule)
+          schema = pa.schema(schema_obj.children_schema)
+          yield schema.serialize().to_pybytes()
+        if arrow_array and arrow_schema:
```
Comment (oliviawebber) [inline src/genomicsdb.pyx:375]: The `and arrow_schema` part of this check seems redundant given the `arrow_schema == NULL` check just above, or is there some truthiness conversion?

File: src/genomicsdb.pyx (hunk 6 above)
Comment (nalinigans) [reply src/genomicsdb.pyx:369]: We can get to the schema only when arrow_array() returns the first time. I could add another separate set of semaphores for arrow_schema to wait on if necessary. See `genomicsdb_nanoarrow_processor.cc` in https://github.com/GenomicsDB/GenomicsDB/pull/338.

File: src/genomicsdb.pyx (hunk 7 above)
Comment (nalinigans) [reply src/genomicsdb.pyx:375]: Agreed!

File: requirements.txt (hunk 5 above)
Comment (nalinigans) [reply requirements.txt:31]: @oliviawebber , any suggestions for `pyarrow` versions? I don't want to introduce incompatibilities with other packages.

Comment (oliviawebber) [reply requirements.txt:31]: If there's no known minimum, my usual rule of thumb is to pick a version from like ~1 year ago which would give `14.0.0`. The one library I know we use that might eventually pull in `pyarrow` looks like it requires `>= 7.0.0`.

Comment (mlathara) [reply requirements.txt:31]: Sorta related: I ran into this issue https://stackoverflow.com/questions/78634235/numpy-dtype-size-changed-may-indicate-binary-incompatibility-expected-96-from recently. Based on that (and maybe this migration guide https://numpy.org/devdocs/numpy_2_0_migration_guide.html) do we also need to pin to `numpy<2.0.0`?

**Review by mlathara — COMMENTED** [review]
Some minor questions

File: requirements.txt (hunk 5 above)
Comment (nalinigans) [reply requirements.txt:31]: Yes, I am using numpy C interface that may need porting to 2. So will have
```
numpy>=1.25,<2.0.0
pyarrow>=14.0.0
```

File: src/genomicsdb.pyx [hunk 8]
```diff
@@ -305,6 +312,78 @@ cdef class _GenomicsDB:
...
+          array_obj = _ArrowArrayWrapper._import_from_c_capsule(schema_capsule, array_capsule)
+          arrays = list()
+          for i in range(array_obj.n_children):
+            arrays.append(pa.array(array_obj.child(i)))
+          yield pa.RecordBatch.from_arrays(arrays, schema=schema).serialize().to_pybytes()
+        else:
+          break
+
+      if non_blocking:
+        query_thread.join()
```
[the remark this answers is not in the document]
Comment (nalinigans) [reply src/genomicsdb.pyx:385]: The client does the cleanup here as the operations are `zero copy`. See https://github.com/GenomicsDB/GenomicsDB-Python/blob/4e3717033e278f5b9169bb4721204906bcd7938e/src/utils.pxi#L102 and https://github.com/GenomicsDB/GenomicsDB-Python/blob/4e3717033e278f5b9169bb4721204906bcd7938e/src/utils.pxi#L105 where we register the callbacks for cleanup for schema and array respectively.

**Review by mlathara — APPROVED** [review]
lgtm

**Pull request merged**